flink-connector-examples/src/main/java/io/pravega/example/flink/streamcuts/process/StreamBookmarker.java

```diff
@@ -150,46 +145,50 @@ public void open(Configuration parameters) {
      */
     @Override
     public void processElement(Tuple2<Integer, Double> value, Context ctx, Collector<SensorStreamSlice> out) throws IOException {
-        if (lastValue.value() != null) return;
-        // Infer the slice of the stream for which events are < 0.
-        if (value.f1 < 0 && startStreamCut.value() == null) {
-            startStreamCut.update(getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM)));
-            LOG.warn("Start bookmarking a stream slice at: {} for sensor {}.", startStreamCut.value(), value.f0);
-        } else if (value.f1 >= 0 && startStreamCut.value() != null && lastStreamCutBeforeSlice.value() == null) {
-            lastStreamCutBeforeSlice.update(getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM)));
-            lastValue.update(value);
-            long sliceCreationTime = ctx.timerService().currentProcessingTime() + StreamBookmarker.CHECKPOINT_INTERVAL;
-            LOG.warn("Found events > 0 for sensor {}, waiting for the next stream cut update to create the slice at {}.", value.f0, sliceCreationTime);
-            ctx.timerService().registerProcessingTimeTimer(sliceCreationTime);
+        // Initialize task state for pending slices.
+        if (pendingBookmarks.value() == null) {
+            pendingBookmarks.update(new HashMap<>());
+        }
+
+        if (value.f1 < 0 && !pendingBookmarks.value().containsKey(value.f0)) {
```

Comment:
You do not need a map here, because the contract of `ValueState` is that Flink tracks a separate value per key (and that's why a `ValueStateDescriptor` may be applied only to a `KeyedStream` (i.e. `.keyBy(0)`).   Therefore your map will always contain at most one value.

Per the [docs](https://ci.apache.org/projects/flink/flink-docs-release-1.6/dev/stream/state/state.html#keyed-state):

> ValueState<T>: This keeps a value that can be updated and retrieved (scoped to key of the input element as mentioned above, so there will possibly be one value for each key that the operation sees). The value can be set using update(T) and retrieved using T value().

Reply:
You are right, I changed that. Thanks.